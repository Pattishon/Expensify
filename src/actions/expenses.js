import uuid from "uuid";

export const addExpense = (
  expense = {
    description: "",
    note: "",
    amount: 0,
    createdAt: 0
  }
) => ({
  type: "ADD_EXPENSE",
  expense: {
    id: uuid(),
    ...expense
  }
});

export const removeExpense = id => ({
  type: "REMOVE_EXPENSE",
  id
});

export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});
