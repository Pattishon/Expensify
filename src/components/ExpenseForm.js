import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

const now = moment();

class ExpenseForm extends React.Component {
  // state = {
  //   description: "",
  //   amount: "",
  //   note: "",
  //   createdAt: moment(),
  //   calendarFocused: false,
  //   error: ""
  // };

  // componentDidMount() {
  //   if (this.props.expense) {
  //     console.log(this.props.expense);
  //     this.setState(() => ({
  //       ...this.state,
  //       ...this.props.expense//przeparsowaÄ‡
  //     }));
  //   }
  // }
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ""
    };
  }

  handleDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  handleNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  handleAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  //calendar handlers
  handleDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  handleFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  //submit handler
  handleSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: "Please provide description and amount" }));
    } else {
      const { description, amount, note, createdAt } = this.state;
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description,
        amount: parseFloat(amount, 10) * 100,
        note,
        createdAt: createdAt.valueOf()
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.handleSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          className="text-input"
          type="text"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.handleDescriptionChange}
        />
        <input
          className="text-input"
          type="text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.handleAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.handleDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.handleFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className="textarea"
          placeholder="Add a note for the expense (optional)"
          value={this.state.note}
          onChange={this.handleNoteChange}
        ></textarea>
        <div>
          <button className="button">
            {this.props.expense ? "Save changes" : "Add Expense"}
          </button>
        </div>
      </form>
    );
  }
}

export default ExpenseForm;
