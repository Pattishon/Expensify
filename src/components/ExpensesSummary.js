import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import selectExpenses from "../selectors/expenses";
import getExpensesTotal from "../selectors/expenses-total";

export const ExpensesSummary = props => {
  return (
    <div>
      <h4>
        Viewing {props.expensesCount}{" "}
        {props.expensesCount === 1 ? "expense" : "expenses"} totalling{" "}
        {numeral(props.expensesTotal / 100).format("$0,0.00")}
      </h4>
    </div>
  );
};

const mapStateToProps = state => ({
  expensesCount: selectExpenses(state.expenses, state.filters).length,
  expensesTotal: getExpensesTotal(selectExpenses(state.expenses, state.filters))
});

export default connect(mapStateToProps)(ExpensesSummary);
