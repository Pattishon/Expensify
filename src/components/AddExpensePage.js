import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";

export const AddExpensePage = props => {
  const handleSubmit = expense => {
    props.startAddExpense(expense);
    props.history.push("/");
  };
  return (
    <div>
      <div className="page-header">
        <div className="content-container">
          <h2 className="page-header__title">Add Expense</h2>
        </div>
      </div>
      <div className="content-container">
        <ExpenseForm onSubmit={handleSubmit} />
      </div>
    </div>
  );
};

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(
  null,
  mapDispatchToProps
)(AddExpensePage);
