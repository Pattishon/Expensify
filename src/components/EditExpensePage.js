import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export const EditExpensePage = props => {
  const handleSubmit = expense => {
    props.startEditExpense(props.expense.id, expense);
    props.history.push("/");
  };
  const handleRemove = () => {
    props.startRemoveExpense(props.expense.id);
    props.history.push("/");
  };
  return (
    <div>
      <div className="page-header">
        <div className="content-container">
          <h2 className="page-header__title">Edit Expense</h2>
        </div>
      </div>
      <div className="content-container">
        <ExpenseForm onSubmit={handleSubmit} expense={props.expense} />
        <button className="button button--secondary" onClick={handleRemove}>
          Remove Expense
        </button>
      </div>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};
const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: id => dispatch(startRemoveExpense(id))
});
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
