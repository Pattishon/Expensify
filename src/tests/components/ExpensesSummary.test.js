import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";
import expenses from "../fixtures/expenses";

test("should render ExpensesSummary for 1 expense correctly", () => {
  const wrapper = shallow(
    <ExpensesSummary expensesCount={1} expensesTotal={123} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpensesSummary for multiple expenses correctly", () => {
  const wrapper = shallow(
    <ExpensesSummary expensesCount={23} expensesTotal={123342.41} />
  );
  expect(wrapper).toMatchSnapshot();
});
