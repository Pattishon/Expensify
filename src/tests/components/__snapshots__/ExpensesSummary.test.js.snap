// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary for 1 expense correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h2
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling
       
      <span>
        $1.23
      </span>
    </h2>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Create Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary for multiple expenses correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h2
      className="page-header__title"
    >
      Viewing 
      <span>
        23
      </span>
       
      expenses
       totalling
       
      <span>
        $1,233.42
      </span>
    </h2>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Create Expense
      </Link>
    </div>
  </div>
</div>
`;
